Server 1978 helm reg doc updates (#7154)
circleci/circleci-docs: jekyll/_cci2/server/installation/migrate-from-server-3-to-server-4.adoc

** link:https://github.com/mikefarah/yq#install[yq]
** link:https://github.com/helm/helm#install[helm]
** link:https://github.com/databus23/helm-diff#install[helm-diff]
** link:https://docs.microsoft.com/en-us/cli/azure/install-azure-cli[Azure CLI]

[#migration]
== Migration

> To perform an in-place upgrade of CircleCI Server from v3.x to v4.x, you will execute a migration script, which will perform the following tasks:

Reviewer: ```suggestion
To perform an in-place upgrade of CircleCI server from v3.x to v4.x, you will execute a migration script, which will perform the following tasks:
```

* Export the KOTS config values into a yaml file.
* Modify the exported YAML file to be used with `helm` to migrate to server v4.x.
* Annotate all the kubernetes resources for `helm`.
* Execute postgres db `domain` migration.
* Execute cleanup steps which will delete all the KOTS related resources.
* Display an output message with the next steps.

> IMPORTANT: In CircleCI Server 4.0, we have changed the way we store workflow data. Previously we had stored your workflow data in Postgres however we've moved to store this data in your storage bucket (GCS/S3/Minio). After migrating, there will be a period where your workflow data will be unavailable. This is due to your existing workflow data being transfered to your storage bucket. The time taken depends on how much existing workflow data you have, however you may scale the workflow pods that are responsible for the migration to speed up the process. You may scale as much as your cluster will allow. In later steps we will show you how to scale your pods.

Reviewer 2: ```suggestion
IMPORTANT: In CircleCI Server 4.0, we have changed the way we store workflow data. Previously we had stored your workflow data in Postgres; however, we have moved this data into your storage bucket (GCS/S3/Minio). After migrating, there will be a period where your workflow data will be unavailable. This is due to your existing workflow data being transferred to your storage bucket. The time taken depends on how much existing workflow data you have; however, you may scale the workflow-conductor pods that are responsible for the migration to speed up the process. You may scale as much as your cluster will allow. In later steps we will show you how to scale your pods.
```

> IMPORTANT: In CircleCI Server 4.0, we have changed the way we store workflow data. Previously we had stored your workflow data in Postgres; however, we have moved this data into your storage bucket (GCS/S3/Minio). After migrating, there will be a period where your workflow data will be unavailable. This is due to your existing workflow data being transferred to your storage bucket. The time taken depends on how much existing workflow data you have; however, you may scale the workflow-conductor pods that are responsible for the migration to speed up the process. You may scale as much as your cluster will allow. In later steps we will show you how to scale your pods.

Reviewer: ```suggestion
IMPORTANT: In CircleCI server v4.x, we have changed the way we store workflow data. Previously we had stored your workflow data in Postgres. We have moved this data into your storage bucket (GCS/S3/MinIO). After migrating, there will be a period where your workflow data will be unavailable. This is due to your existing workflow data being transferred to your storage bucket. The time taken depends on how much existing workflow data you have, however, you may scale the `workflow-conductor` pods that are responsible for the migration to speed up the process. You may scale as much as your cluster will allow. In later steps we will show you how to scale your pods.
```

* Display the output message (`-f message`)
* Rerun the PostgreSQL database migration (`-f flyway`)

[#create-docker-secret-for-circleci-image-registry]
=== 1. Create docker secret for CircleCI image registry
The migration script requires access to the CircleCI image registry to run the migration function. The image registry is also used in the server v4.x installation. Create the `docker-registry` secret in the same Kubernetes namespace in-which CircleCI server is installed. Image registry credentials will be provided by your CircleCI point of contact. 

In a terminal run the following, substituting all values indicated by `< >`:

[source,shell]
----
kubectl -n <namespace> create secret docker-registry regcred \

>   --docker-server=cciserver.azurecr.io \

Reviewer 2: That's a good point, I always have it prefixed with the protocol when I configure a cluster

>   --docker-server=cciserver.azurecr.io \

Reviewer 2: For safety,
```suggestion
  --docker-server=https://cciserver.azurecr.io \
```

# Collect your postgres user's password and the PVC attached to your postgres instance
export POSTGRESQL_PASSWORD=$(kubectl get secret --namespace <namespace> postgresql -o jsonpath="{.data.postgresql-password}" | base64 --decode)
export POSTGRESQL_PVC=$(kubectl get pvc --namespace <namespace> -l app.kubernetes.io/instance=circleci-server,role=primary -o jsonpath="{.items[0].metadata.name}")
# remove the postgres statefulset without terminating your postgres instance
kubectl delete statefulsets.apps postgresql --namespace <namespace> --cascade=orphan
# remove the existing secret containing your postgres password. This will get recreated during upgrade.
kubectl delete secret postgresql --namespace <namespace>
-------------------------------------------------------------------------
## Helm Diff
The Helm Diff tool is used to verify that the changes between your current install and the upgrade are expected

# diff command
helm diff upgrade <release-name> -n <namespace> -f <path>/kots-exporter/output/helm-values.yaml --show-secrets --context 5 oci://cciserver.azurecr.io/circleci-server --version 4.0.0
-------------------------------------------------------------------------

> ## Helm Upgrade CircleCI Server

Reviewer 2: ```suggestion
----

## Helm Upgrade CircleCI Server
```

helm upgrade <release-name> -n <namespace> -f <path>/kots-exporter/output/helm-values.yaml oci://cciserver.azurecr.io/circleci-server --version <version-to-upgrade-to> --force

NOTE: After server 3.x to 4.x migration, You must rerun the Nomad terraform with modified value of 'server_endpoint' variable
It should be - <domain-name>:4647

> ----

Reviewer: @soulchips I think this code block ending needs moving above the note?

Reviewer 2: This is actually correct as-is https://github.com/CircleCI-Public/server-scripts/blob/main/kots-exporter/kots-exporter.sh#L377-L378

> ----

Reviewer 2: ```suggestion
```

[source,shell]
----
############ Upgrade Commands ################
Before upgrading to 4.0, follow below steps:

Your helm values file with your installation config is here:
- <path>/kots-exporter/output/helm-values.yaml
-------------------------------------------------------------------------

> ## Progres Chart Upgrade Preparation

Reviewer: ```suggestion
## Postgres chart upgrade preparation
```

Author: this is as the output from the command we run

Reviewer: Oh right sorry!

> Upgrading to server CircleCI Server 4.0 includes upgrading the Posgres chart

Reviewer 2: ```suggestion
Upgrading to server CircleCI server 4.0 includes upgrading the Posgres chart
```

> Upgrading to server CircleCI server 4.0 includes upgrading the Posgres chart

Reviewer: ```suggestion
Upgrading to CircleCI server v4.x includes upgrading the Postgres chart.
```

> Before upgrading, we need to prepare your postgres instance.

Reviewer: ```suggestion
Before upgrading, we need to prepare your Postgres instance.
```

> This is only needed if your Postgres Instance is not externalized.

Reviewer: ```suggestion
This is only needed if your Postgres instance is not externalized.
```

> # Collect your postgres user's password and the PVC attached to your postgres instance

Reviewer: ```suggestion
[source,shell]
----
# Collect your Postgres user's password and the PVC attached to your Postgres instance
```

Author: this is part of the output of the migration command

export POSTGRESQL_PASSWORD=$(kubectl get secret --namespace <namespace> postgresql -o jsonpath="{.data.postgresql-password}" | base64 --decode)
export POSTGRESQL_PVC=$(kubectl get pvc --namespace <namespace> -l app.kubernetes.io/instance=circleci-server,role=primary -o jsonpath="{.items[0].metadata.name}")
# remove the postgres statefulset without terminating your postgres instance
kubectl delete statefulsets.apps postgresql --namespace <namespace> --cascade=orphan
# remove the existing secret containing your postgres password. This will get recreated during upgrade.
kubectl delete secret postgresql --namespace <namespace>

> -------------------------------------------------------------------------

Reviewer: ```suggestion
-------------------------------------------------------------------------
----

```

Author: this is also part of the output of the migration command

## Helm Diff
The Helm Diff tool is used to verify that the changes between your current install and the upgrade are expected

> # diff command
> helm diff upgrade <release-name> -n <namespace> -f <path>/kots-exporter/output/helm-values.yaml --show-secrets --context 5 oci://cciserver.azurecr.io/circleci-server --version 4.0.0
> -------------------------------------------------------------------------

Reviewer: ```suggestion
[source,shell]
----
# diff command
helm diff upgrade <release-name> -n <namespace> -f <path>/kots-exporter/output/helm-values.yaml --show-secrets --context 5 oci://cciserver.azurecr.io/circleci-server --version 4.0.0
-------------------------------------------------------------------------
----
```

Author: this is also part of the output of the migration command

## Helm Upgrade CircleCI Server

> 

Reviewer: ```suggestion
[source,shell]
```

Author: this is also part of the output of the migration command

> helm upgrade <release-name> -n <namespace> -f <path>/kots-exporter/output/helm-values.yaml oci://cciserver.azurecr.io/circleci-server --version 4.0.0 --force

Reviewer: ```suggestion
helm upgrade <release-name> -n <namespace> -f <path>/kots-exporter/output/helm-values.yaml oci://cciserver.azurecr.io/circleci-server --version <version-to-upgrade-to> --force
```

Reviewer: Would this be ok to make this work for future releases?

----

NOTE: After server 3.x to 4.x migration, You must rerun the Nomad terraform with modified value of 'server_endpoint' variable
It should be - <domain-name>:4647

> IMPORTANT: Please not if your Postgres instance is not externalized. When upgrading to CircleCI Server 4.0, the Postgres chart is also being upgraded.

Reviewer 2: ```suggestion
IMPORTANT: Please note if your Postgres instance is not externalized. When upgrading to CircleCI Server 4.0, the Postgres chart is also being upgraded.
```

> IMPORTANT: Please note if your Postgres instance is not externalized. When upgrading to CircleCI Server 4.0, the Postgres chart is also being upgraded.

Reviewer: ```suggestion
IMPORTANT: If your Postgres instance is not externalized, when upgrading to CircleCI Server v4.0, the Postgres chart is also being upgraded.
```

Please note the command needed before running helm upgrade above. 

[step-3-validate-your-helm-value-file]
=== Step 3 - Validate your helm-value file

> The migration script provides the path your custom helm-values.yaml with your CircleCI server config. Check and validate the generated helm value file and modify if necessary.

Reviewer: I can't quite follow this sentence, could you check for a missing word?
